client/scripts.js

```diff
@@ -1,22 +1,263 @@
+// Function to fetch topics from the backend and display them
+async function fetchTopics() {
+  try {
+    const query = await fetch('http://localhost:8080/api/v1/topic');
+    const topics = await query.json();
+    displayTopics(topics);
+  } catch (error) {
+    console.error('Error fetching topics:', error);
+  }
+}
+
+// Function to display topics as cards and dropdown links
+function displayTopics(topics) {
+  topics.forEach((topic) => {
+    const path = topic.name.toLowerCase().split(' ').join('-');
+
+    // Create the topic card
+    const card = document.createElement('li');
+    card.className = 'card';
+
+    // Add event listener for topic card click
+    card.addEventListener('click', () => handleTopicClick(path));
+
+    card.innerHTML = topic.name;
+    cards.appendChild(card);
+
+    // Create the dropdown link
+    const link = document.createElement('a');
+    link.textContent = topic.name;
+    link.href = '#';
+
+    // Add event listener for dropdown link click
+    link.addEventListener('click', (e) => {
+      e.preventDefault();
+      handleTopicClick(path);
+    });
+
+    dropdown.appendChild(link);
   });
-  card.innerHTML = topic.name;
-  cards.appendChild(card);
+}
+
+// Function to handle topic card or dropdown click
+function handleTopicClick(path) {
+  const token = localStorage.getItem('token');
+  if (token) {
+    // If user is already logged in, navigate to the topic page
+    window.open(`pages/${path}/index.html`, '_self');
+  } else {
+    // Show login popup if the user is not logged in
+    popup.style.display = 'block';
+    lastClickedPath = path; // Store the clicked path
+  }
+}
+
+// Function to display messages (error/success)
+function showMessage(message, type = 'error') {
+  messageContainer.style.display = 'flex';
+  authMessage.style.display = 'block';
+  authMessage.textContent = message;
+  authMessage.style.color = type === 'error' ? 'red' : 'green';
+}
+
+// Function to hide messages
+function hideMessage() {
+  messageContainer.style.display = 'none';
+  authMessage.style.display = 'none';
+  authMessage.textContent = '';
+}
+
+// Email validation function
+function validateEmail(email) {
+  const re = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
+  return re.test(email);
+}
+
+// Password validation function (must contain one number, one capital letter, and one special character)
+function validatePassword(password) {
+  const re = /^(?=.*[A-Z])(?=.*\d)(?=.*[!@#$%^&*()_+]).{8,}$/;
+  return re.test(password);
+}
 
-  const link = document.createElement('a');
-  link.textContent = topic.name;
-  link.setAttribute('href', `pages/${path}/index.html`);
-  dropdown.appendChild(link);
+// Close the popup when the close button is clicked
+closePopup.addEventListener('click', () => {
+  popup.style.display = 'none';
 });
+
+// // close popup when clicking outside the popup
+// window.addEventListener('click', (e) => {
+//   if (e.target === popup) {
+//     popup.style.display = 'none';
+//   }
+// });
```

Comment:
Consider using a multi-line comment if you intend to comment out more than a couple of lines (and you think the comment should stay in this case).

```suggestion
/* // close popup when clicking outside the popup
window.addEventListener('click', (e) => {
   if (e.target === popup) {
     popup.style.display = 'none';
   }
}); */
```